### PR TITLE
ledgerutil: Add "verify" command

### PR DESCRIPTION
**Pull request by shimos**
#### Type of change

- New feature

#### Description

We would like to propose a new command in Ledger Utility that checks the integrity of a single ledger. The existing commands in Ledger Utility aim at checking the ledger when any difference is found by comparing two ledgers. The purpose of the command proposed is to complement these commands by finding any defect in ledger files in a single peer that could be caused by potential software bugs or disk errors.

This patch adds a new command, “verify,” to Ledger Utility, which performs integrity checks for a specified ledger. Currently, the checks for the hash values in the headers of the blocks are implemented.

A use case assumed is to run the command locally in a peer periodically, which can be utilized as one of the health checks for the peer.

#### Additional details

This patch contains a sample ledger that is tweaked to have a hash value error (the hash value for Block 0 is modified not to match the block payload). Using the ledger, the command proposed can detect the error as follows:

```
$ ./build/bin/ledgerutil verify ./internal/ledgerutil/testdata/sample_bad_ledger -o /tmp/ledgerutil-verify-result-bad
...
Successfully executed verify tool. Some error(s) are found.
$ cat /tmp/ledgerutil-verify-result-bad/mychannel_verification_result/blocks.json
[
{"number":0,"valid":false,"errors":["DataHash mismatch"]}
,
{"number":1,"valid":false,"errors":["PreviousHash mismatch"]}
]
```

### CODE REVIEW

File: internal/ledgerutil/verify/verify.go
Comment (denyeart) [inline internal/ledgerutil/verify/verify.go]:
```suggestion
// VerifyLedger verifies the integrity of ledgers in a single peer
```
Exported function comment should start with the exact function name.

Also, please add more comment details about what is actually verified.

File: internal/ledgerutil/verify/verify.go
Comment (denyeart) [inline internal/ledgerutil/verify/verify.go]: It is good practice to use Go zero values, for a boolean it would be 'false'.

Then the variable could be called something like ledgerErrorFound.... it would remain false and then turn true if an error is found.

File: internal/ledgerutil/verify/verify.go
Comment (denyeart) [inline internal/ledgerutil/verify/verify.go]: These can start with lower case to signal unexported.

File: internal/ledgerutil/verify/verify.go
Comment (denyeart) [inline internal/ledgerutil/verify/verify.go]:
```suggestion
// checkBlock checks a block's hash
```

Not as big a deal for unexported functions, but still the function comment should start with the function name.

**Review by denyeart — COMMENTED** [review]
This is a great addition to ledgerutil, thanks!

**Comment by shimos** [comment]
@denyeart Thank you for the review! I have fixed the patch according to your suggestions.

**Comment by denyeart** [comment]
@manish-sethi Could you review as well? Anything else you think should be in a ledger 'verify' utility?

File: internal/ledgerutil/verify/verify.go
Comment (manish-sethi) [inline internal/ledgerutil/verify/verify.go]: If a channel has been bootstapped by a snapshot, the starting block would be a non-zero. You can use the `info` struct in line 75 above to determine this, and get the first available block number.

**Review by manish-sethi — COMMENTED** [review]
Not to discourage, but personally, I do not see much value (usefulness) in this command. Still, I won't object from merging this if any other maintainer wants (and @denyeart likes this) as it is in the offline utility space. Left a comment below.

**Comment by denyeart** [comment]
@shimos One of the new unit tests failed.

**Comment by shimos** [comment]
@manish-sethi Thank you for the comment and review. I have added a logic to determine the first block available and a unit test using a ledger bootstrapped from a snapshot.

@denyeart Thank you for letting me know. I have fixed the failed unit test.

File: internal/ledgerutil/verify/verify.go
Comment (denyeart) [inline internal/ledgerutil/verify/verify.go]: The casual user may not know what "number" is.
The other utility uses the json property "blockNum" for this, would be good to be consistent.

**Review by denyeart — COMMENTED** [review]
Sorry it took me a while... I wanted to try out the user experience before merging.

The user experience is good. I've just highlighted one suggestion and then I think it will be ready for merge.

We'll also need to add the documentation, do you want to do that in this PR or a next one?

See the existing doc at:
https://hyperledger-fabric.readthedocs.io/en/latest/commands/ledgerutil.html

There are instructions for how to add CLI doc here:
https://github.com/hyperledger/fabric/blob/main/docs/source/docs_guide.md#adding-a-new-cli-command

**Comment by shimos** [comment]
@denyeart Thank you for trying out and the suggestion. I've addressed your point.

> We'll also need to add the documentation, do you want to do that in this PR or a next one?

Yes, I'm happy to work on the documentation and submit another PR.

**Review by denyeart — APPROVED** [review]

**Comment by denyeart** [comment]
@shimos Ok, I'll merge this PR and you can work on the doc in a next PR. Thanks!

**Comment by denyeart** [comment]
@Mergifyio backport release-2.5

**Pull request merged**
